### PR TITLE
Adds on z impact interaction for SM.

### PR DESCRIPTION
**Pull request by Helg2**
## About The Pull Request
Adds interaction for supermatter crystal when it falls on something/someone.
I changed intercept_z_fall line that prevents sm dusting itself from continue to return because it didn't spin/dust on fall.
## Why It's Good For The Game
Presumably a little bit more chaos.
## Changelog
:cl:
add: SM crystal can now dust someone or something if it falls on it.
/:cl:

### CODE REVIEW

File: code/datums/components/supermatter_crystal.dm [hunk 1]
```diff
@@ -218,11 +220,34 @@
...
 	return FALL_INTERCEPTED | FALL_NO_MESSAGE
 
+/datum/component/supermatter_crystal/proc/on_z_impact(datum/source, turf/impacted_turf, levels)
+	SIGNAL_HANDLER
+
+	for(var/atom/movable/hit_object as anything in impacted_turf)
+		if(parent == hit_object)
+			return
+
+		var/mob/living/poor_target = locate(/mob/living) in impacted_turf
```
Comment (MrMelbert) [inline code/datums/components/supermatter_crystal.dm:236]: Why do you iterate over all movables in the turf just to do a `locate(mob) in`

Comment (Helg2) [reply code/datums/components/supermatter_crystal.dm:236]: Well, it didn't work properly from what I remember, so I did this, I'll recheck it later.

Comment (MrMelbert) [reply code/datums/components/supermatter_crystal.dm:236]: `for(var/mob/living/guy in impacted_turf)`?

Comment (Helg2) [reply code/datums/components/supermatter_crystal.dm]: Thanks

File: code/datums/components/supermatter_crystal.dm [hunk 2]
```diff
@@ -218,11 +220,33 @@
...
+/datum/component/supermatter_crystal/proc/on_z_impact(datum/source, turf/impacted_turf, levels)
+	SIGNAL_HANDLER
+
+	for(var/atom/movable/hit_object as anything in impacted_turf)
+		if(parent == hit_object)
+			return
+
+		var/atom/atom_source = source
+
+		for(var/mob/living/poor_target in impacted_turf)
```
Comment (MrMelbert) [inline code/datums/components/supermatter_crystal.dm:238]: You seemed to have misunderstood something

Delete all of this
```dm
for(var/atom/movable/hit_object as anything in impacted_turf)
		if(parent == hit_object)
			return

		var/atom/atom_source = source
```
Right now you're iterating over every atom in the turf, just so you can iterate over ever mob in the turf for every atom there is

Just have
```dm
	for(var/mob/living/poor_target in impacted_turf)
		...
```

**Review by MrMelbert — CHANGES_REQUESTED** [review]

**Comment by Colovorat** [comment]
instead of dropping anvil on people we will be dropping supermatter

File: code/datums/components/supermatter_crystal.dm [hunk 3]
```diff
@@ -218,11 +220,34 @@
...
+
+	for(var/atom/movable/hit_object as anything in impacted_turf)
+		if(parent == hit_object)
+			return
+
+		if(!iseffect(hit_object))
+			consume(atom_source, hit_object)
+			playsound(get_turf(atom_source), 'sound/effects/supermatter.ogg', 50, TRUE)
+			atom_source.visible_message(span_danger("\The [atom_source] smacks into \the [hit_object] out of nowhere and rapidly flashes to ash."), null,
+				span_hear("You hear a loud crack as you are washed with a wave of heat."))
```
Comment (Jacquerel) [inline code/datums/components/supermatter_crystal.dm:249]:
```suggestion
		if(iseffect(hit_object))
			continue
		consume(atom_source, hit_object)
		playsound(get_turf(atom_source), 'sound/effects/supermatter.ogg', 50, TRUE)
		atom_source.visible_message(span_danger("\The [atom_source] smacks into \the [hit_object] out of nowhere and rapidly flashes to ash."), null,
			span_hear("You hear a loud crack as you are washed with a wave of heat."))
```
Prefer early returns when the alternative is several indented lines

**Review by Jacquerel — CHANGES_REQUESTED** [review]
If you drop this on a tile containing a mob and several objects it's going to result in some chat spam of several fairly long messages. Maybe keep the full output for mobs it falls on (only spams if you drop it on a pile of mice or something) but condense it to one message if it eats several objects at once? The specifics aren't necessarily so important in that case.

**Comment by Helg2** [comment]
> If you drop this on a tile containing a mob and several objects it's going to result in some chat spam of several fairly long messages. Maybe keep the full output for mobs it falls on (only spams if you drop it on a pile of mice or something) but condense it to one message if it eats several objects at once? The specifics aren't necessarily so important in that case.

I mean i can change the message to just `\The [atom_source] smacks into whatever there was out of nowhere and rapidly flashes to ash.`
It's the only thing that comes to mind.

**Comment by Jacquerel** [comment]
> > If you drop this on a tile containing a mob and several objects it's going to result in some chat spam of several fairly long messages. Maybe keep the full output for mobs it falls on (only spams if you drop it on a pile of mice or something) but condense it to one message if it eats several objects at once? The specifics aren't necessarily so important in that case.
> 
> I mean i can change the message to just `\The [atom_source] smacks into whatever there was out of nowhere and rapidly flashes to ash.` It's the only thing that comes to mind.

Keep a counter of how many objects it consumed and then display a generic message if there's one or more. What you've written will do.
`\the [atom_source], smacks into the plating out of nowhere, reducing everything below to ash.`

**Comment by Helg2** [comment]
I lack the knowledge how to make the message properly, so I'll do it later i guess.

**Pull request merged**